### PR TITLE
SNOW-655614 Data type validation: Boolean

### PR DESCRIPTION
**Pull request by sfc-gh-lsembera**
This is the first commit in a series of commits focusing on comprehensive data type validation. It establishes testing infrastructure and adds tests for BOOLEAN data type.

It also fixes a problem that negative integers were previously evaluated as false.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -2,4 +2,4 @@ repos:
 - hooks:
   - id: secret-scanner
   repo: git@github.com:snowflakedb/casec_precommit.git
-  rev: v1.5
+  rev: v1.8
```
Comment (sfc-gh-lsembera) [inline .pre-commit-config.yaml:5]: With the old version of the secret-scanner, git commit on my machine is failing with

```
/home/lsembera/.cache/pre-commit/repo47p9zrdc/secret_scanner.sh: line 48: pip: command not found 
/home/lsembera/.cache/pre-commit/repo47p9zrdc/secret_scanner.sh: line 22: detect-secrets: command not found 
/home/lsembera/.cache/pre-commit/repo47p9zrdc/secret_scanner.sh: line 23: [: !=: unary operator expected
```

File: src/test/java/net/snowflake/ingest/TestUtils.java [hunk 2]
```diff
@@ -222,8 +225,10 @@ public static Connection getConnection() throws Exception {
    * @throws Exception
    */
   public static Connection getConnection(boolean isStreamingConnection) throws Exception {
-    if (!isStreamingConnection && snowpipeConn != null) return snowpipeConn;
-    if (isStreamingConnection && streamingConn != null) return streamingConn;
+    if (!isStreamingConnection && snowpipeConn != null && !streamingConn.isClosed())
+      return snowpipeConn;
+    if (isStreamingConnection && streamingConn != null && !streamingConn.isClosed())
+      return streamingConn;
```
Comment (sfc-gh-tjones) [inline src/test/java/net/snowflake/ingest/TestUtils.java:231]: Nit, let's wrap these in `{` and `}`

File: src/test/java/net/snowflake/ingest/TestUtils.java [hunk 3]
```diff
@@ -317,4 +322,27 @@ public static SimpleIngestManager getManagerUsingBuilderPattern(
...
+   * Given a channel and expected offset, this method waits until the last committed offset is equal
+   * to the passed offset
+   */
+  public static void waitForOffset(SnowflakeStreamingIngestChannel channel, String expectedOffset)
+      throws InterruptedException {
+    int counter = 0;
+    String lastCommittedOffset = null;
+    while (counter < 600) {
+      String currentOffset = channel.getLatestCommittedOffsetToken();
+      if (Objects.equals(expectedOffset, currentOffset)) {
```
Comment (sfc-gh-tjones) [inline src/test/java/net/snowflake/ingest/TestUtils.java:336]: Can we just not do a regular string comparison here?

File: src/test/java/net/snowflake/ingest/TestUtils.java [hunk 4]
```diff
@@ -317,4 +322,27 @@ public static SimpleIngestManager getManagerUsingBuilderPattern(
         .setUserAgentSuffix(userAgentSuffix)
         .build();
   }
+
+  /**
+   * Given a channel and expected offset, this method waits until the last committed offset is equal
```
Comment (sfc-gh-tjones) [inline src/test/java/net/snowflake/ingest/TestUtils.java:327]: Let's document how long we intend to wait.

File: src/test/java/net/snowflake/ingest/TestUtils.java [hunk 5]
```diff
@@ -317,4 +322,27 @@ public static SimpleIngestManager getManagerUsingBuilderPattern(
...
+    int counter = 0;
+    String lastCommittedOffset = null;
+    while (counter < 600) {
+      String currentOffset = channel.getLatestCommittedOffsetToken();
+      if (Objects.equals(expectedOffset, currentOffset)) {
+        return;
+      }
+      lastCommittedOffset = currentOffset;
+      counter++;
+      Thread.sleep(100);
```
Comment (sfc-gh-tjones) [inline src/test/java/net/snowflake/ingest/TestUtils.java:341]: We might want to bump this up to a full second

**Review by sfc-gh-tjones — APPROVED** [review]
Nice! Thanks for separating this out from your other PR!

File: src/test/java/net/snowflake/ingest/TestUtils.java (hunk 5 above)
Comment (sfc-gh-lsembera) [reply src/test/java/net/snowflake/ingest/TestUtils.java:341]: Increasing the sleep to 1 second would prolog the test run significantly. This method is called many times and the commit is usually processed quickly, so waiting for 1 second in each iteration is unnecessary.

File: src/test/java/net/snowflake/ingest/TestUtils.java (hunk 3 above)
Comment (sfc-gh-lsembera) [reply src/test/java/net/snowflake/ingest/TestUtils.java]: Fixed

File: src/test/java/net/snowflake/ingest/TestUtils.java (hunk 4 above)
Comment (sfc-gh-lsembera) [reply src/test/java/net/snowflake/ingest/TestUtils.java]: Fixed

File: src/test/java/net/snowflake/ingest/TestUtils.java (hunk 2 above)
Comment (sfc-gh-lsembera) [reply src/test/java/net/snowflake/ingest/TestUtils.java]: Fixed

**Review by sfc-gh-tjones — APPROVED** [review]

File: src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java [hunk 6]
```diff
@@ -691,4 +692,29 @@ static boolean convertStringToBoolean(String value) {
...
+        String.format("Value cannot be converted to %s", type));
+  }
+
+  /**
+   * Before passing a value to an exception string, we should limit how many characters are
+   * displayed. Important especially for "value exceeds max column size" exceptions.
+   *
+   * @param value Value to adapt for exception message
+   */
+  private static String sanitizeValueForExceptionMessage(Object value) {
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java:715]: I'm not sure if this is really needed, what if customer need to copy the value?

Comment (sfc-gh-lsembera) [reply src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java:715]: You mean copy from the logs? We can increase the limit, but I think there should be some limit. Previously if somebody tried to ingest a 100MB string, SDK rejects it, but dump those 100MB into the logs.

**Pull request merged**